File: schemas/content-moderation/drizzle/moderator_performance.ts
// moderator_performance: Pre-aggregated moderator performance metrics.
// See README.md for full design rationale.

import {
  pgTable,
  uuid,
  integer,
  numeric,
  timestamp,
  unique,
  index,
} from "drizzle-orm/pg-core";
import { users } from "../../auth-rbac/drizzle/users";

export const moderatorPerformance = pgTable(
  "moderator_performance",
  {
    id: uuid("id").primaryKey().defaultRandom(),

    moderatorId: uuid("moderator_id")
      .notNull()
      .references(() => users.id, { onDelete: "cascade" }),

    periodStart: timestamp("period_start", { withTimezone: true }).notNull(),
    periodEnd: timestamp("period_end", { withTimezone: true }).notNull(),

    itemsReviewed: integer("items_reviewed").notNull().default(0),
    itemsActioned: integer("items_actioned").notNull().default(0),
    averageReviewTimeSeconds: integer("average_review_time_seconds")
      .notNull()
      .default(0),
    appealsOverturned: integer("appeals_overturned").notNull().default(0),
    accuracyScore: numeric("accuracy_score").notNull().default("1.0"),

    computedAt: timestamp("computed_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
  },
  (table) => [
    unique("uq_moderator_performance_moderator_period").on(
      table.moderatorId,
      table.periodStart,
      table.periodEnd
    ),
    index("idx_moderator_performance_period").on(
      table.periodStart,
      table.periodEnd
    ),
  ]
);

File: schemas/content-moderation/spacetimedb/moderator_performance.rs
// moderator_performance: Pre-aggregated moderator performance metrics.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = moderator_performance, public)]
pub struct ModeratorPerformance {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub moderator_id: String, // FK → users.id (cascade delete)

    #[index(btree)]
    pub period_start: Timestamp,
    #[index(btree)]
    pub period_end: Timestamp,
    // Composite unique: (moderator_id, period_start, period_end) — enforce in reducer logic

    pub items_reviewed: i32,
    pub items_actioned: i32,
    pub average_review_time_seconds: i32,
    pub appeals_overturned: i32,
    pub accuracy_score: f64,

    pub computed_at: Timestamp,
    pub created_at: Timestamp,
}
